skimage/restoration/_nl_means_denoising.pyx

```diff
@@ -0,0 +1,565 @@
+
+    Parameters
+    ----------
+    image: ndarray
+        3-D input data to be denoised
+    s: int, optional
+        size of patches used for denoising
+    d: int, optional
+        maximal distance in pixels where to search patches used for denoising
+    h: float, optional
+        cut-off distance (in gray levels). The higher h, the more permissive
+        one is in accepting patches.
+    """
+    if s % 2 == 0:
+        s += 1  # odd value for symmetric patch
+    cdef int offset = s / 2
+    # Image padding: we need to account for patch size, possible shift,
+    # + 1 for the boundary effects in finite differences
+    cdef int pad_size = offset + d + 1
+    cdef DTYPE_t [:, :, ::1] padded = np.ascontiguousarray(util.pad(image,
+                                pad_size, mode='reflect').astype(np.float32))
+    cdef DTYPE_t [:, :, ::1] result = np.zeros_like(padded)
+    cdef DTYPE_t [:, :, ::1] weights = np.zeros_like(padded)
+    cdef DTYPE_t [:, :, ::1] integral = np.zeros_like(padded)
+    cdef int n_pln, n_row, n_col, t_pln, t_row, t_col, \
+             x_pln, x_row, x_col
+    cdef int x_pln_integral_min, x_pln_integral_max, \
+             x_row_integral_min, x_row_integral_max, \
+             x_col_integral_min, x_col_integral_max
+    cdef int x_pln_dist_min, x_pln_dist_max, x_row_dist_min, x_row_dist_max, \
+             x_col_dist_min, x_col_dist_max
+    cdef float weight, distance
+    cdef float alpha
+    cdef float h_square = h ** 2.
+    cdef float s_cube = s ** 3.
+    cdef float s_cube_h_square = h_square * s_cube
+    n_pln, n_row, n_col = image.shape
+    n_pln += 2 * pad_size
+    n_row += 2 * pad_size
+    n_col += 2 * pad_size
+    # Outer loops on patch shifts
+    # With t2 >= 0, reference patch is always on the left of test patch
+    for t_pln in range(-d, d + 1):
+        x_pln_integral_min = max(1, - t_pln)
+        x_pln_integral_max = min(n_pln, n_pln - t_pln)
+        x_pln_dist_min = max(offset, offset - t_pln)
+        x_pln_dist_max = min(n_pln - offset, n_pln - offset - t_pln)
+        for t_row in range(-d, d + 1):
+            x_row_integral_min = max(1, - t_row)
+            x_row_integral_max = min(n_row, n_row - t_row)
+            x_row_dist_min = max(offset, offset - t_row)
+            x_row_dist_max = min(n_row - offset, n_row - offset - t_row)
+            for t_col in range(0, d + 1):
+                x_col_integral_min = max(1, - t_col)
+                x_col_integral_max = min(n_col, n_col - t_col)
+                x_col_dist_min = max(offset, offset - t_col)
+                x_col_dist_max = min(n_col - offset, n_col - offset - t_col)
+                # alpha is to account for patches on the same column
+                # distance is computed twice in this case
+                if t_col == 0 and (t_pln is not 0 or t_row is not 0):
+                    alpha = 0.5
+                else:
+                    alpha = 1.
+                integral = np.zeros_like(padded)
```

Comment:
yes, integral turns out to be the same array as 

```
integral_image((padded[t_row:, t_col, :] - padded[:-t_row, :-t_col])**2)
```

Thanks for noticing this, I was implementing the formula provided in the article of Froment to compute ``integral``, and then I wasn't thinking much about it. 

It will remove a lot of code that is hard to read if I use ``integral_image``. However, it would also use more memory if ``(padded[t_row:, t_col, :] - padded[:-t_row, :-t_col])**2`` is computed -- which is not a problem for most images, but might be a problem for large 3D images. Well, I can inline a custom integral function with this part of the code, and write a test to check whether the result is the same as for ``ìntegral_image``. Unless you have a better idea...